webapp/src/ts/services/db-sync.service.ts

```diff
@@ -166,31 +168,76 @@ export class DBSyncService {
       telemetryEntry.recordSuccess(result);
     } catch (err) {
       telemetryEntry.recordFailure(err, this.knownOnlineState);
-      console.error(`Error replicating from remote server`, err);
+      console.error('Error replicating from remote server', err);
       return err;
     }
   }
 
+  private getCurrentSeq(): Promise<number> {
+    return this.dbService.get().info().then(info => info.update_seq);
+  }
 
-  private getCurrentSeq() {
-    return this.dbService.get().info().then(info => info.update_seq + '');
+  private getLastReplicatedSeq(): number {
+    return Number(window.localStorage.getItem(LAST_REPLICATED_SEQ_KEY)) || 0;
   }
 
-  private getLastReplicatedSeq() {
-    return window.localStorage.getItem(LAST_REPLICATED_SEQ_KEY);
+  private setLastReplicatedSeq(sequence: number) {
+    if (!sequence) {
+      return;
+    }
+
+    const lastReplicatedSeq = this.getLastReplicatedSeq();
+    if (lastReplicatedSeq > sequence) {
+      return;
+    }
+
+    window.localStorage.setItem(LAST_REPLICATED_SEQ_KEY, sequence.toString());
   }
 
   private getLastReplicationDate() {
     return window.localStorage.getItem(LAST_REPLICATED_DATE_KEY);
   }
 
+  private async makeBidirectionalReplication(replicateFromServer: boolean, successiveSyncs = 0) {
+    const replicationErrors = {
+      to: await this.replicateTo(),
+      from: replicateFromServer ? await this.replicateFrom() : null,
+    };
+    const hasErrors = replicationErrors.to || replicationErrors.from;
+    let syncState = await this.getSyncState(hasErrors);
+
+    const isSyncRequired = syncState.to === SyncStatus.Required || syncState.from === SyncStatus.Required;
+    if (isSyncRequired && successiveSyncs < MAX_SUCCESSIVE_SYNCS) {
+      successiveSyncs += 1;
+      syncState = await this.makeBidirectionalReplication(replicateFromServer, successiveSyncs);
+    }
+
+    return syncState;
+  }
+
+  private async getSyncState(hasErrors): Promise<SyncState> {
+    const currentSeq = await this.getCurrentSeq();
+    const lastReplicatedSeq = this.getLastReplicatedSeq();
+
+    if (!hasErrors && (!this.canReplicateToServer || currentSeq === lastReplicatedSeq)) {
+      return { to: SyncStatus.Success, from: SyncStatus.Success };
+    }
+
+    if (hasErrors && currentSeq === lastReplicatedSeq) {
+      // No changes to send, but may have some to receive
+      return { state: SyncStatus.Unknown };
+    }
+
+    // Definitely need to sync something
+    return { to: SyncStatus.Required, from: SyncStatus.Required };
```

Comment:
For this specific ticket, I thought of including a new status for when there are no errors and the current seq is higher than the last replicated ( `if (!hasErrors && (!this.canReplicateToServer || currentSeq > lastReplicatedSeq))` ) to trigger a successive sync. But it was kind of the same as `SyncStatus.Required`, then for simplicity, I decided to leave it like this.